crates/typst/src/eval/plugin.rs

```diff
@@ -0,0 +1,59 @@
+use std::ops::Deref;
+use std::sync::Mutex;
+
+use crate::diag::SourceResult;
+use host_wasmer::PluginInstance;
+use typst::diag::At;
+
+#[derive(Debug)]
+pub struct Plugin(pub Mutex<PluginInstance>);
+
+impl PartialEq for Plugin {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.lock().unwrap().deref() == other.0.lock().unwrap().deref()
```

Comment:
Regarding `Hash` and `PartialEq`:

I think it would be good have a constructor `Plugin::new(data: Bytes) -> StrResult<Self>` and then store `data` as a field next to the `instance`. Then PartialEq and Hash can just compare and hash the `data` (which is very cheap since `Bytes` is prehashed).

This assumes that two plugins created from the same source code are always equal, i.e. that the plugin can't have mutable state. But if that's not the case we have a problem anyway because of memoization.

Reply:
that's more reasonable indeed, on it